### PR TITLE
Allow value in patch remove operation

### PR DESCRIPTION
**Pull request by deepak-cyral**
## Description of the change

This PR fixes two issues.

- Allow value field to be specified in a remove operation in a patch request - this is needed since some implementations do not use the filter in path for this purpose.
- Deletion of an array element was broken since deletion in the middle of iteration on array elements changed the array indices. This is fixed as well.

## Type of change
- [X] Bug fix (non-breaking change that fixes an issue)
- [X] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

### Development

- [X] Lint rules pass locally
- [X] The code changed/added as part of this pull request has been covered with tests
- [X] All tests related to the changed code pass in development

### Code review 

- [X] This pull request has a descriptive title and information useful to a reviewer. There may be a screenshot or screencast attached
- [ ] Jira issue referenced in commit message and/or PR title

### Testing

Added unit tests covering the new and changed code.

Additionally, these changes were tested with OneLogin, which sends resource IDs as values in PATCH operations. In order to test this, The following test was performed successfully:

- Some users were removed from a group, triggering a successful PATCH remove operation on user and group resources in the SCIM user and group databases
- Some user information was updated, triggering a successful PATCH update operation on user resources in the SCIM user and group databases
- Some user information was removed, triggering a successful PATCH remove operation on user resources in the SCIM user and group databases
- Some user information was added, triggering a successful PATCH update operation on user resources in the SCIM user and group databases

### CODE REVIEW

File: pkg/v2/crud/traverse.go [hunk 1]
```diff
@@ -79,10 +81,32 @@ func (t traverser) traverseSelectedElements(query *expr.Expression) error {
 
 func (t traverser) traverseQualifiedElements(filter *expr.Expression) error {
 	return t.nav.ForEachChild(func(index int, child prop.Property) error {
-		t.nav.At(index)
-		if err := t.nav.Error(); err != nil {
-			return err
+		// index is not trustworthy if an earlier child has been deleted by
```
Comment (deepak-cyral) [inline pkg/v2/crud/traverse.go:84]: This change is necessary to allow dealing with deletion of an array element as the array elements are being visited to see if they pass the filter.

File: pkg/v2/crud/crud.go [hunk 2]
```diff
@@ -53,10 +56,30 @@ func Delete(resource *prop.Resource, path string) error {
 	if err != nil {
 		return err
 	}
+	var query *expr.Expression
 
-	return defaultTraverse(resource.RootProperty(), skipMainSchemaNamespace(resource, head), func(nav prop.Navigator) error {
-		return nav.Delete().Error()
-	})
+	if value != nil {
```
Comment (deepak-cyral) [inline pkg/v2/crud/crud.go:61]: At a high level, what's happening here is that we form a filter expression from the given value(s) and append it to the expression created from the given path. This effectively augments the path with the a filter condition that says that the element value must match _one_ of the supplied values.

File: pkg/v2/crud/expr/expression.go [hunk 3]
```diff
@@ -105,6 +105,20 @@ func (e *Expression) ContainsFilter() bool {
 	return false
 }
 
+// Append appends a given expression to the tail of this expression.
+func (e *Expression) Append(expr *Expression) *Expression {
+	if e == nil {
+		return expr
+	}
+	head := e
```
Comment (deepak-cyral) [inline pkg/v2/crud/expr/expression.go:113]: The `head` variable is redundant and can be removed easily.

File: pkg/v2/crud/expr/from-value.go [hunk 4]
```diff
@@ -0,0 +1,109 @@
+package expr
+
+import (
+	"fmt"
+)
+
+func FromValueList(list []interface{}) (*Expression, error) {
```
Comment (deepak-cyral) [inline pkg/v2/crud/expr/from-value.go:7]: add godoc comment.

File: pkg/v2/crud/expr/from-value.go [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+	if len(list) == 0 {
+		return nil, nil
+	}
+	var exprs []*Expression
+	for _, v := range list {
+		value, ok := v.(map[string]interface{})
+		if !ok {
+			return nil, fmt.Errorf("invalid value type for creating expression: %T", v)
```
Comment (deepak-cyral) [inline pkg/v2/crud/expr/from-value.go:15]: embed spec.ErrInvalidSyntax?

File: pkg/v2/crud/crud.go [hunk 6]
```diff
@@ -53,10 +56,30 @@ func Delete(resource *prop.Resource, path string) error {
...
+		switch v := value.(type) {
+		case map[string]interface{}:
+			query, err = expr.FromValue(v)
+		case []interface{}:
+			query, err = expr.FromValueList(v)
+		default:
+			return fmt.Errorf("values (%v) for a delete operation is of unsupported type %T",
```
Comment (deepak-cyral) [inline pkg/v2/crud/crud.go:68]: embed ErrInvalidSyntax

File: pkg/v2/crud/expr/from-value.go [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+			subexprs, err := getLeafExpressions(prefix, value)
+			if err != nil {
+				return nil, err
+			}
+			exprs = append(exprs, subexprs...)
+		}
+		return exprs, nil
+	default:
+		// cannot handle arrays, anything else is unexpected.
+		return nil, fmt.Errorf("unexpected value %v", v)
```
Comment (deepak-cyral) [inline pkg/v2/crud/expr/from-value.go:107]: invalid syntax

**Review by hbaackmann — APPROVED** [review]
These changes were tested with OneLogin, which sends resource IDs as values in PATCH operations. In order to test this, The following test was performed successfully:

- Some users were removed from a group, triggering a successful PATCH remove operation on user and group resources in the SCIM user and group databases
- Some user information was updated, triggering a successful PATCH update operation on user resources in the SCIM user and group databases
- Some user information was removed, triggering a successful PATCH remove operation on user resources in the SCIM user and group databases
- Some user information was added, triggering a successful PATCH update operation on user resources in the SCIM user and group databases

**Pull request merged**